src/auth_context.rs

```diff
@@ -0,0 +1,155 @@
+            if p.is_null() {
+                None
+            } else {
+                Some(CStr::from_ptr(p).to_str().expect("valid UTF-8 data"))
+            }
+        }
+    }
+
+    /// `true` if the client has provided a valid certificate (or other auth method
+    /// considered valid by gRPC).
+    /// `false` in non-secure scenarios.
+    pub fn peer_is_authenticated(&self) -> bool {
+        assert!(!self.ctx.is_null());
+        unsafe { grpc_sys::grpc_auth_context_peer_is_authenticated(self.ctx) != 0 }
+    }
+
+    /// `AuthContext[peer_identity_property_name()]`
+    ///
+    /// There may be several of them (for instance if `x509_subject_alternative_name` is selected)
+    pub fn peer_identity(&self) -> AuthPropertyIter {
+        assert!(!self.ctx.is_null());
+        unsafe {
+            // grpc_auth_context_peer_identity returns empty_iterator when self.ctx is NULL
+            let iter = grpc_sys::grpc_auth_context_peer_identity(self.ctx);
+            AuthPropertyIter {
+                iter,
+                _lifetime: PhantomData,
+            }
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a AuthContext {
+    type Item = AuthProperty<'a>;
+    type IntoIter = AuthPropertyIter<'a>;
+
+    /// Iterate over the AuthContext properties
+    fn into_iter(self) -> Self::IntoIter {
+        unsafe {
+            // grpc_auth_context_property_iterator returns empty_iterator when self.ctx is NULL
+            let iter = grpc_sys::grpc_auth_context_property_iterator(self.ctx);
+            AuthPropertyIter {
+                iter,
+                _lifetime: PhantomData,
+            }
+        }
+    }
+}
+
+impl Drop for AuthContext {
+    fn drop(&mut self) {
+        unsafe { grpc_sys::grpc_auth_context_release(self.ctx) }
+    }
+}
+
+pub struct AuthPropertyIter<'a> {
+    iter: grpc_auth_property_iterator,
+    _lifetime: PhantomData<&'a grpc_auth_property_iterator>,
+}
+
+impl<'a> Iterator for AuthPropertyIter<'a> {
+    type Item = AuthProperty<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // grpc_auth_property_iterator_next returns empty_iterator when self.iter is NULL
+        let prop = unsafe { grpc_sys::grpc_auth_property_iterator_next(&mut self.iter) };
+        if prop.is_null() {
+            None
+        } else {
+            Some(AuthProperty {
+                prop,
+                _lifetime: PhantomData,
+            })
+        }
+    }
+}
+
+/// Auth properties are elements of the AuthContext. They have a name
+/// (a key of type string) and a value which can be a string or binary data.
+pub struct AuthProperty<'a> {
+    prop: *const grpc_auth_property,
+    _lifetime: PhantomData<&'a grpc_auth_property>,
+}
+
+impl<'a> AuthProperty<'a> {
+    pub fn name(&self) -> &'a str {
+        unsafe {
+            CStr::from_ptr((*self.prop).name)
+                .to_str()
+                .expect("Auth property name should be valid UTF-8 data")
```

Comment:
Let's return result. Should avoid panic in library.

Reply:
Looking at https://grpc.github.io/grpc/core/md_doc_server_side_auth.html it seems these properties are stored with names that are hardcoded by the gRPC Core team, and seem to be UTF-8 so it seems to me that these could be safely unwrapped.

Reply:
https://github.com/grpc/grpc/blob/2b200d2313e739fa1841edf3ea86ab8532862505/include/grpc/grpc_security_constants.h#L26-L34

Reply:
https://github.com/grpc/grpc/blob/2b200d2313e739fa1841edf3ea86ab8532862505/src/core/lib/security/security_connector/ssl_utils.cc#L221-L271

Reply:
LGTM, you can just use `from_utf8_unchecked`.

Reply:
Isn't that a bit dangerous if they ever decide to add properties that have dynamic names based on the content of certificates ?

Reply:
I don't think they would do that as the design doc says it should be a string already: https://github.com/grpc/grpc/blob/master/doc/server_side_auth.md#authproperty. But I'm not sure whether an interceptor can add an invalid string. Let's keep the check then.